targets/TARGET_NORDIC/TARGET_NRF5x/flash_api.c

```diff
@@ -38,82 +38,167 @@
 
 #if DEVICE_FLASH
 
-#include "flash_api.h"
-#include "nrf_nvmc.h"
-#include "nrf_soc.h"
+#include "hal/flash_api.h"
+#include "hal/lp_ticker_api.h"
+
+#include "nrf_fstorage.h"
 
 #if defined(SOFTDEVICE_PRESENT)
-#include "nrf_sdm.h"
+#include "nrf_fstorage_sd.h"
+#else
+#include "nrf_fstorage_nvmc.h"
 #endif
 
+#include <stdbool.h>
+
+#define PAGE_ERASE_TIMEOUT_US (200 * 1000)  // Max. value from datasheet: 89.7 ms
+
+NRF_FSTORAGE_DEF(static nrf_fstorage_t nordic_fstorage) = { 0 };
+
 int32_t flash_init(flash_t *obj)
 {
     (void)(obj);
+
+    ret_code_t result = NRF_SUCCESS;
+
+    /* Only initialize once. */
+    static bool do_init = true;
+
+    if (do_init) {
+        do_init = false;
+
+        /* Set instance to cover the whole flash. */
+        nordic_fstorage.p_flash_info    = NULL;
+        nordic_fstorage.evt_handler     = NULL;
+        nordic_fstorage.start_addr      = 0;
+        nordic_fstorage.end_addr        = NRF_FICR->CODESIZE * NRF_FICR->CODEPAGESIZE;
+
+        /* Initialize either SoftDevice API or NVMC API. 
+         * SoftDevice API should work both when the SoftDevice is enabled or disabled.
+         * NVMC API is used when the SoftDevice is not present.
+         */
 #if defined(SOFTDEVICE_PRESENT)
-    uint8_t sd_enabled;
-    if ((sd_softdevice_is_enabled(&sd_enabled) == NRF_SUCCESS) && sd_enabled == 1) {
-        return -1;
-    }
+        result = nrf_fstorage_init(&nordic_fstorage, &nrf_fstorage_sd, NULL);
+#else
+        result = nrf_fstorage_init(&nordic_fstorage, &nrf_fstorage_nvmc, NULL);
 #endif
-    return 0;
+
+        /* Initialize low power ticker for timeouts. */
+        lp_ticker_init();
+    }
+
+    /* Convert Nordic SDK error code to mbed HAL. */
+    return (result == NRF_SUCCESS) ? 0 : -1;
 }
 
 int32_t flash_free(flash_t *obj)
 {
     (void)(obj);
+
     return 0;
 }
 
 int32_t flash_erase_sector(flash_t *obj, uint32_t address)
 {
     (void)(obj);
-#if defined(SOFTDEVICE_PRESENT)
-    uint8_t sd_enabled;
-    if ((sd_softdevice_is_enabled(&sd_enabled) == NRF_SUCCESS) && sd_enabled == 1) {
-        return -1;
+
+    ret_code_t result = NRF_ERROR_NO_MEM;
+
+    /* Setup stop watch for timeout. */
+    uint32_t start_us = lp_ticker_read();
+    uint32_t now_us = start_us;
+
+    /* Retry if flash is busy until timeout is reached. */
+    while (((now_us - start_us) < PAGE_ERASE_TIMEOUT_US) &&
+           (result == NRF_ERROR_NO_MEM)) {
```

Comment:
isn't while loop breaks after first try because result will be different then NRF_ERROR_NO_MEM? or it doesn't change in case of timeout??

Reply:
nrf_fstorage_erase will queue up the erase operation if there is space in the fifo or return NRF_ERROR_NO_MEM. We only retry if the fifo is full and only for a fixed amount of time.

If the call returns any other error we will return -1 to the caller and 0 on success.